### PR TITLE
Add an example API

### LINKED ISSUE #57: Provide hooks to W3C SFrame transform
https://w3c.github.io/webrtc-insertable-streams/#sframe defines a native SFrame transform.
It would be good to define hooks in this spec that could be referred from https://w3c.github.io/webrtc-insertable-streams/#sframe.
I can see the following hooks that might be worth exposing:
- Encrypting/decrypting a frame
- Setting keys
- Ratcheting
- Error reporting, especially on decoding side: decryption failure, missing key...

### PR DESCRIPTION
**Pull request by bifurcation**
Fixes #57

### CODE REVIEW

File: draft-ietf-sframe-enc.md [hunk 1]
```diff
@@ -875,6 +875,70 @@ The authors wish to specially thank Dr. Alex Gouaillard as one of the early
...
+
+This section describes a notional API that an SFrame implementation might
+expose.  The core concept is an "SFrame context", within which KID values are
+meaningful.  In the key management scheme described in {{sender-keys}}, each
+sender has a different context; in the scheme described in {{mls}}, all senders
+share the same context.
+
+An SFrame context stores mappings from KID values to "key contexts", which are
+different depending on whether the KID is to be used for sending or receiving
+(an SFrame keys should never be used for both operations).  A key context tracks
```
Comment (GhostofCookie) [inline draft-ietf-sframe-enc.md:890]:
```suggestion
(an SFrame key should never be used for both operations).  A key context tracks
```
Or possibly 

```suggestion
(SFrame keys should never be used for both operations).  A key context tracks
```
But likely the first suggestion.

File: draft-ietf-sframe-enc.md [hunk 2]
```diff
@@ -875,6 +875,70 @@ The authors wish to specially thank Dr. Alex Gouaillard as one of the early
...
+the key and salt associated to the KID, and the current CTR value.  A key
```
Comment (GhostofCookie) [inline draft-ietf-sframe-enc.md:891]:
```suggestion
the key and salt associated to the KID, and the current CTR value.  The key
```
It makes sense as is. Whichever you prefer.

File: draft-ietf-sframe-enc.md [hunk 3]
```diff
@@ -875,6 +875,70 @@ The authors wish to specially thank Dr. Alex Gouaillard as one of the early
...
+* **Create an SFrame context:** The context is initialized with a ciphersuite and
+  no KID mappings.
+* **Adding a key for sending:** The key and salt are derived from the base key, and
+  used to initialize a send context, together with a zero counter value.
+* **Adding a key for receiving:** The key and salt are derived from the base key, and
+  used to initialize a send context.
+* **Encrypt a plaintext:** Encrypt a given plaintext using the key for a given KID,
+  including the specified metadata.
+* **Decrypt an SFrame ciphertext:** Decrypt an SFrame ciphertext with the KID
+  and CTR values specified in the SFrame Header, and the provided metadata.
```
Comment (GhostofCookie) [inline draft-ietf-sframe-enc.md:905]: Just a markdown question here; Should the bullets be - instead of *? Or is this an RFC standard thing?

Comment (bifurcation) [reply draft-ietf-sframe-enc.md:905]: Markdown doesn't care, nor does the IETF.

File: draft-ietf-sframe-enc.md (hunk 2 above)
Comment (bifurcation) [reply draft-ietf-sframe-enc.md:891]: Nah, this is "A context to be" as in "Any context which is to be..."

File: draft-ietf-sframe-enc.md [hunk 4]
```diff
@@ -875,6 +875,70 @@ The authors wish to specially thank Dr. Alex Gouaillard as one of the early
...
+  cipher_suite: CipherSuite,
+  send_keys: HashMap<KeyId, SendKeyContext>,
+  recv_keys: HashMap<KeyId, RecvKeyContext>,
+}
+
+trait SFrameContextMethods {
+  fn create(cipher_suite: CipherSuite) -> Self;
+  fn add_send_key(&self, kid: KeyId, base_key: &[u8]);
+  fn add_recv_key(&self, kid: KeyId, base_key: &[u8]);
+  fn encrypt(&mut self, kid: KeyId, metadata: &[u8], plaintext: &[u8]) -> Vec<u8>;
```
Comment (suhasHere) [inline draft-ietf-sframe-enc.md:936]: should we add a note on what forms AAD in this setup ?

**Review by suhasHere — APPROVED** [review]
minor clarification question

LGTM

File: draft-ietf-sframe-enc.md (hunk 4 above)
Comment (bifurcation) [reply draft-ietf-sframe-enc.md:936]: No, the AAD is constructed internal to `encrypt()` / `decrypt()` according to the SFrame rules.  This is just capturing the outer API exposed to the application.

**Pull request merged**